### PR TITLE
Offer improvements

### PR DESCRIPTION
**Pull request by slipeer**
- Empty name error handling
- Setup name from LDAP on user registration
- Update user name on each login (because user name may be updated in LDAP)
- Readme improvement

### CODE REVIEW

**Comment by slipeer** [comment]
Now it update mail from LDAP too. But i suggest use mail from LDAP only with _LDAP ready_ Identity Server like [mxisd](https://github.com/kamax-io/mxisd).

**Comment by slipeer** [comment]
`samAccountName` has modern alternative: `userPrincipalName` but in most systems that i view still used samAccountName too.

`givenName` in example - it's my mistake i will fix it now.

Also this weekend I found that checking the presence of an attribute value is not entirely correct. Now I will fix this too.

About the tests - this is the most difficult. I registered [Issue in the synapse to extend _AccountHandler proxy object methods](https://github.com/matrix-org/synapse/issues/2085). When this is done, we can update code to use new methods and fix the tests.

**Comment by slipeer** [comment]
@mweinelt test now work properly. Without `twisted.trial`... clean `unittest` - it's better documented.

**Comment by mweinelt** [comment]
@erikjohnston Any opinion on the test changes? You've written most of that, so it would take me some time to dig in, which I currently cannot spend.

If test changes are fine this is good to merge.

**Comment by erikjohnston** [comment]
What was the rational of redoing the tests?

**Comment by slipeer** [comment]
@erikjohnston 

1. add to the account_handler the stubs of the functions required for the new functionality
2. added a new test: verification of user authorization with unpresent additional attributes values
3. old tests with new twisted began to emerge error Reactor not clean (I suppose it caused a multiple LDAP server creation)

I can only extend `account_handler` mock object but tests will not work for the third reason.

**Comment by slipeer** [comment]
@erikjohnston @mweinelt 

I made in my repo [branch with old tests](https://github.com/slipeer/matrix-synapse-ldap3/tree/broken_tests) realization to demonstrate [problem with them](https://travis-ci.org/slipeer/matrix-synapse-ldap3/jobs/219561900), which I could not solve.

If anyone can solve it better then I offer, I can bring my master branch to this state.

**Comment by doofy** [comment]
What is preventing this PR from getting merged?

**Comment by alantygel** [comment]
Hi Slipeer, 
In my environment I also need to bind an LDAP user base to synapse. Although your solution already enhances the current behaviour of synapse, it still has some issues.

The first one should be easy to solve: if I change my email address on LDAP, synapse adds another email, instead of really updating it.

However, the most important issue is that data is updated only at login process. If an user changes some parameter, synapse updates only on the next login. And in a mobile environment, where users stay logged in forever, this issue should be definitely tackled. 

I'll start to dig into the code and try to figure out a solution. 

Please tell me what do you think about it.

best

**Comment by slipeer** [comment]
@alantygel about email change: Thanks. This I missed - in the near future I will correct it.
About regular updates users fields from ldap: there no way to do it without changes in matrix-org/synapse
On my server, I solved this problem by blocking the corresponding API on Apache.

It seems that the developers of the project are not very interested in supporting LDAP - changes in this area are considered very slowly.

**Comment by alantygel** [comment]
@slipeer , thanks for the answer.

Could you please detail 

> On my server, I solved this problem by blocking the corresponding API on Apache

?

How about writing changes on synapse to LDAP server? Any ideas on it?

**Comment by maxidor** [comment]
My 2 cents: adding an e-mail/phone number in Riot is just an easy way to do two things:
1. Add an administrative contact for the given user in synapse, only used by that HS, never published anywhere. That's what this authenticator populates.
2. Publish a 3PID association on the configured IS, which is used by other clients to mapping when inviting by e-mail or phone numbers.

Point 2 can be done solely in the IS, and is really what most people care about. Which one are you relying on @alantygel ?

**Comment by slipeer** [comment]
@alantygel about apache - not here write me to @slipeer:kardash.su i try to help, but i filter only displayName user changes
about syncing - no idea. I two months ago proposed small changes https://github.com/matrix-org/synapse/pull/2119 to improve application handler interface  and no one yet looked to it. To synchronize, you need a lot more changes ...

**Comment by alantygel** [comment]
Hi @maxidor ,
In fact, I would already be happy with the solution provided here (Point 1), if : 

- Changes in LDAP server could reflect on synapse without the need of logging in again
- Changes in synapse could write in LDAP base

Additionally, my interest is not on email specifically, but all parameters (displayName, profile image, etc).

Thanks!

**Comment by alantygel** [comment]
@slipeer ,
If I'm not mistaken, emails are still being added and not updated. 

1. Clear user_threepids table;
2. Logging with an LDAP user using matrix_id. User email is entered at user_threepids. 
3. Edit email directly on LDAP
4. Logout/Login -> new email entry at user_threepids is added;

Looking at your code, line 193 only checks if the email exists; there's no update instruction in there.

Trying to get in touch with you via matrix, but I'm not succeeding. My id is @alantygel:matrix.eita.org.br

**Comment by slipeer** [comment]
@alantygel I think I understand you - we thought about different things ;)
I'm afraid there is no way to delete old email automatically when it changed in LDAP - otherwise users will lose ability to bind additional email addresses that are not listed in the LDAP.
In addition, I did not find in synapse any functions that allow you to get email address (3pids) of user.

However, the user can still delete old email through profile properties. I believe that users email addresses in the LDAP are not changing often.

In addition, I realized [Account lockout policy](https://github.com/matrix-org/matrix-synapse-ldap3/pull/26/commits/ef01147e46f947d50e6089076efe01fefd20b568) that helps preventing internal users account lockout (in active directory for example) by bruteforsing password from internet.

File: ldap_auth_provider.py [hunk 1]
```diff
@@ -139,73 +174,102 @@ def check_password(self, user_id, password):
...
-                    try:
-                        mail = attrs[self.ldap_attributes['mail']][0]
-                    except KeyError:
-                        mail = None
-
-                    # create account
+                except:
+                    name = None
+                try:
+                    mail = attrs[self.ldap_attributes['mail']][0]
```
Comment (4nd3r) [inline ldap_auth_provider.py]: what happens when user object has multiple mail attributes? only first one will be used?

Comment (mweinelt) [inline ldap_auth_provider.py]: Indeed that is currently the case. A loop would be appropriate.

**Comment by mweinelt** [comment]
Just an fyi: I'm still waiting for @erikjohnston to give feedback on the tests. I have no idea how they work and too little time to figure them out.

File: ldap_auth_provider.py (hunk 1 above)
Comment (slipeer) [inline ldap_auth_provider.py]: @4nd3r mutliply mail attributes or multi-valued mail attribute with more then one value?

**Comment by slipeer** [comment]
@4nd3r  thanks for tip - fixed in https://github.com/matrix-org/matrix-synapse-ldap3/pull/26/commits/3a5b2ae6e963bd161e8a1f2fef69222dfe2d3072

**Comment by maxidor** [comment]
I checked the code a bit more in depth and I see an issue with registering e-mails at login: the Identity server is NOT contacted, so those will not go through the usual workflow in clients where both the HS and the IS are made aware of the e-mail.

Seems like a big issue to me, as the only way to trigger an IS publication would be to remove and re-add e-mails...

**Comment by slipeer** [comment]
@maxidor  Do you know a Identity Server implementation that can perform a non-interactive binding of an email as 3pid identifier? To me such realization is unknown.

You do not think that synapse will wait while user verify his email address in autentification process? This is at least inconvenient to the user.

Did you see my [comment](https://github.com/matrix-org/matrix-synapse-ldap3/pull/26#issuecomment-288819356) (I added it right after implementing the email from ldap)? Mail from LDAP functional   **only with LDAP ready Identity Server** because of 3pid assotiation must be done on both side: on HS and on IS.

File: ldap_auth_provider.py (hunk 1 above)
Comment (4nd3r) [inline ldap_auth_provider.py]: multiple mail attributes per object.

https://tools.ietf.org/html/rfc4524#section-2.16

**Comment by maxidor** [comment]
@slipeer No Identity server can perform non-interactive binding, because it's up to the user to choose the Identity Server, not the Home Server. It's a mess with the Matrix protocol I think.

About your comment, I did see it, but I am concerned that, *by default*, e-mail addresses are added to the user profile in the HS. There is no guarantee what the IS is, what its capable of, and so you could give a false impression of readiness to the user.

I don't see a good way around it with the current protocol though :(

**Comment by slipeer** [comment]
All functional implemented here:
https://github.com/slipeer/synapse-ldap-password-provider

**Pull request closed without merging**